Constexpr default ctor

vinniefalco opened the pull request:
This makes `error_code` default constructor `constexpr` in C++11 instead of just C++14

vinniefalco commented:
I pushed a fix for defective gcc 4.7.x

pdimov commented:
1. This PR contains unrelated changes and hence can't be merged as-is.
2. What is the code that fails without this PR and passes with it?

pdimov commented on include/boost/system/detail/error_code.hpp:
```diff
@@ -114,7 +114,10 @@ class error_code
     // constructors:
 
-    BOOST_SYSTEM_CONSTEXPR error_code() BOOST_NOEXCEPT:
+#if ! defined(BOOST_GCC) || BOOST_GCC >= 40800
```
BOOST_WORKAROUND would probably be better here.

vinniefalco replied:
Yep, Emil suggested that too but I erred on the side of not adding an additional include. I'm happy to use that macro though.

pdimov replied:
Please do.

vinniefalco replied:
oh dang, system/detail/config.hpp already includes workaround.hpp.

pdimov replied:
Doesn't matter, you should include it here too because it's used.

vinniefalco commented:
1. fixed
2. I have `error_code` as a member of a struct, contained in objects which are constexpr variables with static storage duration:

https://github.com/vinniefalco/http_proto/blob/cbdbb6e959258eebb8970a9b4f1950709d1998b8/include/boost/http_proto/metadata.hpp#L179

(I switched it to the enum as a temporary workaround)

The three "default" values (what you get with a default-constructed object) are constexpr here (`header` contains a `metadata` data member)
https://github.com/vinniefalco/http_proto/blob/cbdbb6e959258eebb8970a9b4f1950709d1998b8/include/boost/http_proto/detail/impl/header.ipp#L73

pdimov commented:
OK, but what is the simplest code example that both demonstrates the failure and is representative of your actual code?

vinniefalco commented:
> OK, but what is the simplest self-contained code example that both demonstrates the failure and is representative of your actual code?

Given this code:

```
#include <boost/system/error_code.hpp>

struct M
{
    constexpr M() = default;

    boost::system::error_code ec;
};

struct H
{
    constexpr H() = default;

    M m;
};

H const& f()
{
    static constexpr H h;
    return h;
}
```

The current tip of develop fails to compile it at all under C++11.

Changing `error_code` in System to `BOOST_CONSTEXPR` will compile under C++11 except for GCC 4.7.x (compiler bug I suppose, I struggled to find a workaround but failed)

Applying this pull request will it compile under C++11, except that it will not be constexpr for GCC 4.7.x. My library does not support anything below 4.8 anyway, so this should be no worse than before my change.

pdimov commented on include/boost/system/detail/error_code.hpp:
```diff
@@ -114,7 +114,10 @@ class error_code
     // constructors:
 
-    BOOST_SYSTEM_CONSTEXPR error_code() BOOST_NOEXCEPT:
+#if ! BOOST_WORKAROUND(BOOST_GCC, >= 40800)
```
Should be <.
